fix is_wp_version_compatible() to accommodate incorrect format x.x.0

afragen opened the pull request:
<!-- Insert a description of your changes here -->
Some users improperly test for releases of the format x.x.0. Unfortunately there is never a WordPress release with this version number and `version_compare( '5.2', '5.2.0', '>=' )` is false.

Test value sent to `is_wp_version_compatible( $required )` to ensure a trailing zero (x.x.0) is fixed to be (x.x) ~and a `wp_trigger_error()` is sent.~

This will correctly test the improper version x.x.0 as x.x.

Trac ticket: https://core.trac.wordpress.org/ticket/59448

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

afragen committed:
fix is_wp_version_compatible() to accommodate incorrect format x.x.0

afragen committed:
wpcs fix

afragen committed:
update test to return wp_trigger_error

afragen committed:
check for string before throwing error

afragen committed:
use E_USER_NOTICE, update test

afragen committed:
refactor check, update $version before wp_trigger_error

afragen committed:
add more tests

afragen committed:
fix substr_count

afragen committed:
use error_log so function doesn't exit

afragen committed:
I think this now works with wp_trigger_error

afragen committed:
Revert "I think this now works with wp_trigger_error"

This reverts commit ea216d5cd648c3ce2ecea0c23283f0f94dd8fbcf.

afragen committed:
more tests

aaronjorbin commented on src/wp-includes/functions.php:
```diff
@@ -8716,6 +8716,16 @@ function is_wp_version_compatible( $required ) {
 	list( $version ) = explode( '-', $wp_version );
 
+	if ( is_string( $required ) && substr_count( $required, '.' ) > 1 && str_ends_with( $required, '.0' ) ) {
+		error_log(
```
This should use `wp_trigger_error`

afragen replied:
I was initially having issues with tests not passing due to the `wp_trigger_error()` message being returned but found a solution by removing the error handler before the test and resetting afterwards.

aaronjorbin commented on src/wp-includes/functions.php:
```diff
@@ -8716,6 +8716,16 @@ function is_wp_version_compatible( $required ) {
+		error_log(
+			__FUNCTION__ . '(): ' .
+			/* translators: s: version string sent to function */
+			sprintf( __( '`%s` Not a valid WordPress version string.' ), $required )
```
It would be helpful to present both what was passed in and what the result was. What do you think of this instead?

"Invalid version 6.0.0 passed to `is_wp_version_compatible`, 6.0 is assumed."

aaronjorbin reviewed (commented):
Great start. I think this might need a test to ensure that the notice is properly raised. I also left a couple of comments on the specific implementation.

afragen committed:
use wp_trigger_error

afragen committed:
Merge branch 'trunk' into 59448

afragen committed:
remover error handler for test that returns wp_trigger_error and reset

afragen committed:
improve version number corrector

afragen committed:
improve warning message

afragen committed:
update related tests to use proper version number

update  to is_wp_version_compatible() sends trigger_warning causing tests to fail.

afragen commented:
Thanks @aaronjorbin, I believe I've fixed the issues you've raised. 

I don't quite understand what you mean by ensuring the notice is properly raised?

aaronjorbin commented:
Thanks, the code changes look great.  Also, glad to hear this helped find another spot where our tests could be improved! 

> I don't quite understand what you mean by ensuring the notice is properly raised

I meant using something like what was implemented in https://core.trac.wordpress.org/changeset/57332 Essentially rather than silencing the error, setting an expect for it to be raised.

afragen committed:
use same exclusion as wpUniquePrefixedId.php:158

afragen commented:
The problem with raising the error to an Exception is that the comparison will then fail and what we've intended is for the provided incorrect version be modified so it's now correct and the comparison continues -- or am I missing something?

aaronjorbin commented:
Sorry I wasn't clear. I pushed some code to your branch since I thought showing the code would help.

By using expectException, the exception won't cause the tests to fail.  I did move the actual call to `is_wp_version_compatible` to it's own line so that the error handler can be restored to it's previous state.

Let me know if this makes sense. It looks like I'll need to clean up some formatting as well, so I can try to add some comments if you think it would help

costdev commented on tests/phpunit/tests/functions/isWpVersionCompatible.php:
```diff
@@ -12,14 +12,34 @@ class Tests_Functions_IsWpVersionCompatible extends WP_UnitTestCase {
-		$this->assertSame( $expected, is_wp_version_compatible( $required ) );
+	public function test_is_wp_version_compatible( $required, $expected, $expect_exception = false ) {
+
+		if ( $expect_exception ) {
```
These latest changes provide a false sense of security, as does the test in [this Core changeset](https://core.trac.wordpress.org/changeset/57332).

For example, if you comment out all datasets _except_ those that should throw an exception, and change the final assertion to `$this->assertSame( $expected . 'should not pass', $result );`, the two datasets will still pass. This is because an exception was correctly detected, but the test also ended as soon as this occurred. This issue also exists with the test in the Core changeset linked above.

Instead, let's consider this approach:

1. Add a new property to the test class:

```php
/**
 * A notice expected during the tests.
 *
 * @var string
 */
private $expected_notice = '';
```
2. Add a `tear_down()` method that resets the property:
```php
/**
 * Resets the `expected_notice` property after each test runs.
 */
public function tear_down() {
	$this->expected_notice = '';
	parent::tear_down();
}
```
3. Adjust the test so that instead of throwing an exception, we note the notice, restore the previous handler, and check that it matches the expected type and string:
```php
public function test_is_wp_version_compatible( $required, $expected, $expect_notice = false ) {
	if ( $expect_notice ) {
		set_error_handler(
			function ( $errno, $errstr ) {
				$this->expected_notice = $errno . ' ' . $errstr;
			},
			E_ALL
		);
	}

	$result = is_wp_version_compatible( $required );

	if ( $expect_notice ) {
		restore_error_handler();
		$this->assertStringStartsWith(
			E_USER_NOTICE . " is_wp_version_compatible(): `$required` Not a valid WordPress version string.",
			$this->expected_notice,
			'The expected notice was not thrown.'
		);
	}

	$this->assertSame( $expected, $result );
}
```

Now, if we change the final assertion to `$this->assertSame( $expected . 'should not pass', $result );`, the two datasets will correctly fail.

Run a few checks locally to see the issue with the existing test, and how the changes above resolve this.

costdev replied:
Looks like the discussion has moved away from having a notice. The above can be disregarded in that case. 🙂

costdev replied:
Un-resolving this as a decision on the notice hasn't finalized.

@aaronjorbin FYI: Take a look at the first comment in this thread regarding an issue with the changeset you referenced and previous commits to this PR that were influenced by it, and an alternative without this issue.

peterwilsoncc commented on tests/phpunit/tests/functions/isWpVersionCompatible.php:
```diff
@@ -12,14 +12,34 @@ class Tests_Functions_IsWpVersionCompatible extends WP_UnitTestCase {
 	 * Tests is_wp_version_compatible().
 	 *
 	 * @dataProvider data_is_wp_version_compatible
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
```
Why are these needed?

costdev replied:
These shouldn't be needed as there are no constants/static variables at play.

afragen replied:
Removed

peterwilsoncc commented on src/wp-includes/functions.php:
```diff
@@ -8735,6 +8735,16 @@ function is_wp_version_compatible( $required ) {
+	if ( is_string( $required ) && substr_count( $required, '.' ) > 1 && str_ends_with( $required, '.0' ) ) {
+		$passed   = $required;
+		$required = preg_replace( '/\.0/', '', $passed, 1 );
+		wp_trigger_error(
+			__FUNCTION__,
+			/* translators: s: version string sent to function */
+			sprintf( __( '`%1$s` Not a valid WordPress version string. `%2$s` is assumed.' ), $passed, $required )
+		);
```
Per notes on ticket -- will need changes to the tests.

afragen replied:
Are we planing on notifying the user in this manner or silently fixing the incorrect version?

peterwilsoncc replied:
Silently fix due to the use of `@since x.x.0` throughout core. 

```suggestion
```

peterwilsoncc commented on tests/phpunit/tests/functions/isWpVersionCompatible.php:
```diff
@@ -85,6 +105,34 @@ public function data_is_wp_version_compatible() {
+				'required' => '5.0',
+				'expected' => true,
+			),
+			'correct version x.0.1'     => array(
+				'required' => '5.0.1',
+				'expected' => true,
+			),
```
Is it worth adding a test to ensure the version is converted correctly? It might require some code shuffling.

costdev replied:
Provided there are tests covering all the bases, testing the output with each input should be enough, and the respective lines can be checked for coverage in a coverage report.

afragen replied:
I added this test to ensure a `.0` in the middle of a version number wouldn’t be changed.

afragen replied:
I will add some new tests for that. I'll have it fixed by tomorrow.

afragen committed:
revert notice and changes to other REST test

afragen committed:
add new test for silent fix

afragen committed:
cleanup dataset

afragen committed:
keep fixed test data for rest-plugins-controller

afragen committed:
re-format

afragen committed:
more cleanup

Commit:
Add notice for incorrect version numbering.

Commit:
Move valid version numbering to the main test.

Commit:
Implement a notice check.

Commit:
Adjust and add datasets to the notice test.

Commit:
F: notice check

afragen committed:
Merge pull request #9 from costdev/is_wp_version_compatible/add_notice_and_test

Add notice and test.

afragen committed:
update translator comment

afragen committed:
refactor function for trailing whitespace add more datasets

azaozz commented:
Thinking this code is far too complex for what it needs to accomplish. We only care to strip `.0` from the end of `$required`, when it is longer than 3 characters, nothing more. So `substr_count()` and `preg_replace()` for a simple string that contains only digits and dots is perhaps "too much"? 

Perhaps something simpler would also work:
```php
if ( is_string( $required ) && strlen( $required ) > 3 && str_ends_with( $required, '.0' ) ) {
    // Strip trailing zero to properly match two digits (major) WP version numbers.
    $required = substr( $required, 0, -2 );
}
```

The `strlen( $required ) > 3` is there to catch versions like 5.0, 6.0, etc. In all other cases the WP versions should not have a trailing zero.

Edit: actually this should probably be `strlen( $required ) > 4` to catch WP versions up to 99.0 (that would still work properly). Seems reasonable to update this code when WP gets to 100.0 :)

azaozz reviewed (changes requested):
See https://github.com/WordPress/wordpress-develop/pull/5677#issuecomment-1958086757.

peterwilsoncc committed:
Remove notice from function.

peterwilsoncc committed:
Update tests for unofficial version numbers to remove notice expectation.

peterwilsoncc committed:
Remove unused variable following 55492561b4d9dbd690a198dd941cb8bbd54e3706

peterwilsoncc committed:
Remove trailing value with substr rather than regex.

peterwilsoncc commented:
@azaozz I've switched `substr` to remove the version number's trailing `.0`

`$required` is developer input that can't be trusted to be entirely correct (the entire point of this ticket ;) so I've kept string count.

peterwilsoncc committed:
Merge branch 'trunk' into 59448

costdev commented on tests/phpunit/tests/functions/isWpVersionCompatible.php:
```diff
@@ -8,6 +8,21 @@
  * @covers ::is_wp_version_compatible
  */
 class Tests_Functions_IsWpVersionCompatible extends WP_UnitTestCase {
+	/**
+	 * A notice expected during the tests.
+	 *
+	 * @var string
+	 */
+	private $expected_notice = '';
+
+	/**
+	 * Resets the `$expected_notice` property after each test runs.
+	 */
+	public function tear_down() {
+		$this->expected_notice = '';
+		parent::tear_down();
+	}
+
```
```suggestion
```

@peterwilsoncc you missed a spot 🧹

peterwilsoncc committed:
Remove tearDown following 55492561b4d9dbd690a198dd941cb8bbd54e3706

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

azaozz reviewed (approved):
> $required is developer input that can't be trusted to be entirely correct...

Right. Thinking it looks better now, and perhaps a very tiny bit faster (afaik PHP needs some time to "initialize" the external lib every time a regexp function is used) :) 
Seems this is good to go if there isn't anything else.

swissspidy commented on tests/phpunit/tests/functions/isWpVersionCompatible.php:
```diff
@@ -43,51 +43,140 @@ public function data_is_wp_version_compatible() {
 	}
 
+	/**
+	 * Tests that is_wp_version_compatible() throws a notice for incorrect version numbering.
```
```suggestion
	 * Tests that is_wp_version_compatible() gracefully handles incorrect version numbering.
```

peterwilsoncc committed:
Fix unit test description following removal of notice.

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

peterwilsoncc committed:
Merge branch 'trunk' into 59448

peterwilsoncc commented:
Merged [r57707](https://core.trac.wordpress.org/changeset/57707) / https://github.com/WordPress/wordpress-develop/commit/2ba8f942cb8e95bf93d19c12f9cd9bb9ae61b72e
